lib/core/src/Cardano/Wallet/Network.hs

```
@@ -251,25 +263,80 @@ data FollowExit
 --
 -- Exits with 'Nothing' in case of error.
 follow
-    :: forall block e. (Show e)
+    :: forall block tr e. (Show e, HasSeverityAnnotation tr)
     => NetworkLayer IO block
     -- ^ The @NetworkLayer@ used to poll for new blocks.
-    -> Tracer IO FollowLog
+    -> Tracer IO (FollowLog tr)
+    -- ^ Logger trace
+    -> IO [BlockHeader]
+    -- ^ A way to get a list of known tips to start from.
+    -- Blocks /after/ the tip will be yielded.
+    -> (NE.NonEmpty block
+        -> BlockHeader
+        -> Tracer IO tr
+        -> IO (FollowAction e))
+    -- ^ Callback with blocks and the current tip of the /node/.
+    -- @follow@ stops polling and terminates if the callback errors.
+    -> (SlotNo -> IO (Either e SlotNo))
+    -- ^ Callback with blocks and the current tip of the /node/.
+    -- @follow@ stops polling and terminates if the callback errors.
+    -> FollowExceptionRecovery
+    -- ^ Whether to recover from exceptions or not.
+    -> (block -> BlockHeader)
+    -- ^ Getter on the abstract 'block' type
+    -> IO ()
+follow nl tr' readCursor forward backward recovery header = do
+    tr <- monitorStatus tr'
+    loop tr
+  where
+    loop tr = do
+        cursor <- readCursor
+        -- TODO: Pool worker would trace MsgStartMonitoring here
+        (follow' nl tr cursor forward header) >>= \act -> do
+            case act of
+                FollowFailure ->
+                    -- TODO: Log here, e.g. MsgCrashMonitoring
+                    case recovery of
```

I would remove FollowFailure and let exceptions bubble up to the caller of `follow`.
The caller can decide whether to retry, possibly using the `recovering` function (which has good logging abilities).

---

As said elsewhere, my goals was to move as much responsibility from the callers to `follow`.

Maybe `recovering` (inside follow) is a good idea, but we didn't use it before at least, so we could keep it separate from this PR.

---

Btw, the pool worker simply stops if you kill and restart the node... We should probably either recover from exceptions for both pool and wallet workers, or re-add an `ExceptT` (#2581)